File: electronicstoreapi/src/main/java/com/company/project/service/OutletService.java
package com.company.project.service;

import java.util.List;

import org.springframework.data.domain.Page;

import com.company.project.dto.request.OutletRequest;
import com.company.project.dto.response.OutletResponse;

public interface OutletService {

    /**
     * Create a new outlet
     * 
     * @param request the outlet data
     * @return the created outlet
     */
    OutletResponse createOutlet(OutletRequest request);

    /**
     * Update an existing outlet
     * 
     * @param id      the outlet id
     * @param request the updated outlet data
     * @return the updated outlet
     */
    OutletResponse updateOutlet(Long id, OutletRequest request);

    /**
     * Get an outlet by id
     * 
     * @param id the outlet id
     * @return the outlet
     */
    OutletResponse getOutletById(Long id);

    /**
     * Delete an outlet
     * 
     * @param id the outlet id
     */
    void deleteOutlet(Long id);

    /**
     * Get all active outlets
     * 
     * @return list of all active outlets
     */
    List<OutletResponse> getAllActiveOutlets();

    /**
     * Get all outlets with pagination
     * 
     * @param page the page number
     * @param size the page size
     * @return paged list of outlets
     */
    Page<OutletResponse> getAllOutlets(int page, int size);

    /**
     * Get outlets by locality id
     * 
     * @param localityId the locality id
     * @return list of outlets in the locality
     */
    List<OutletResponse> getOutletsByLocalityId(Long localityId);

    /**
     * Get outlets by city id
     * 
     * @param cityId the city id
     * @return list of outlets in the city
     */
    List<OutletResponse> getOutletsByCityId(Long cityId);

    /**
     * Get outlets by state id
     * 
     * @param stateId the state id
     * @return list of outlets in the state
     */
    List<OutletResponse> getOutletsByStateId(Long stateId);
}

File: electronicstoreadmin/src/services/outlet.service.ts
import {
    ApiResponse,
    OutletCreateRequest,
    OutletResponse,
    OutletUpdateRequest,
    Page
} from '../types/api-responses';
import { API_ENDPOINTS } from '../utils/api-endpoints';
import { apiFetch } from '../utils/api-fetch';

export class OutletService {
    /**
     * Get all outlets with proper authentication
     * @returns Promise with paginated list of outlets
     */
    static async getAllOutlets(page = 0, size = 10): Promise<ApiResponse<Page<OutletResponse>>> {
        try {
            console.log('OutletService: Calling getAllOutlets API');
            const url = `${API_ENDPOINTS.OUTLETS}?page=${page}&size=${size}`;
            const response = await apiFetch(url);

            if (!response.ok) {
                console.error('OutletService: API returned error status:', response.status);
                throw new Error(`Server returned ${response.status}`);
            }

            const data = await response.json();
            console.log('OutletService: API response data:', data);

            return data;
        } catch (error) {
            console.error('Outlets API call error:', error);
            throw error;
        }
    }

    /**
     * Get active outlets with proper authentication
     * @returns Promise with list of active outlets
     */
    static async getActiveOutlets(): Promise<ApiResponse<OutletResponse[]>> {
        try {
            console.log('OutletService: Calling getActiveOutlets API');
            const response = await apiFetch(API_ENDPOINTS.ACTIVE_OUTLETS);

            if (!response.ok) {
                console.error('OutletService: API returned error status:', response.status);
                throw new Error(`Server returned ${response.status}`);
            }

            const data = await response.json();
            console.log('OutletService: API response data:', data);

            return data;
        } catch (error) {
            console.error('Active outlets API call error:', error);
            throw error;
        }
    }

    /**
     * Get an outlet by ID with proper authentication
     * @param id - Outlet ID
     * @returns Promise with outlet data
     */
    static async getOutletById(id: number): Promise<ApiResponse<OutletResponse>> {
        try {
            const response = await apiFetch(API_ENDPOINTS.OUTLET_BY_ID(id));
            const data = await response.json();
            return data;
        } catch (error) {
            console.error('Outlet API call error:', error);
            throw error;
        }
    }

    /**
     * Create a new outlet with proper authentication
     * @param outlet - Outlet create request data
     * @returns Promise with created outlet data
     */
    static async createOutlet(outlet: OutletCreateRequest): Promise<ApiResponse<OutletResponse>> {
        try {
            const response = await apiFetch(API_ENDPOINTS.OUTLETS, {
                method: 'POST',
                body: JSON.stringify(outlet)
            });
            const data = await response.json();
            return data;
        } catch (error) {
            console.error('Create outlet API call error:', error);
            throw error;
        }
    }

    /**
     * Update an existing outlet with proper authentication
     * @param id - Outlet ID
     * @param outlet - Outlet update request data
     * @returns Promise with updated outlet data
     */
    static async updateOutlet(id: number, outlet: OutletUpdateRequest): Promise<ApiResponse<OutletResponse>> {
        try {
            const response = await apiFetch(API_ENDPOINTS.OUTLET_BY_ID(id), {
                method: 'PUT',
                body: JSON.stringify(outlet)
            });
            const data = await response.json();
            return data;
        } catch (error) {
            console.error('Update outlet API call error:', error);
            throw error;
        }
    }

    /**
     * Delete an outlet with proper authentication
     * @param id - Outlet ID
     * @returns Promise with success/failure response
     */
    static async deleteOutlet(id: number): Promise<ApiResponse<null>> {
        try {
            const response = await apiFetch(API_ENDPOINTS.OUTLET_BY_ID(id), {
                method: 'DELETE'
            });
            const data = await response.json();
            return data;
        } catch (error) {
            console.error('Delete outlet API call error:', error);
            throw error;
        }
    }

    /**
     * Get outlets by locality ID
     * @param localityId - Locality ID
     * @returns Promise with list of outlets
     */
    static async getOutletsByLocality(localityId: number): Promise<ApiResponse<OutletResponse[]>> {
        try {
            const response = await apiFetch(API_ENDPOINTS.OUTLETS_BY_LOCALITY(localityId));
            const data = await response.json();
            return data;
        } catch (error) {
            console.error('Outlets by locality API call error:', error);
            throw error;
        }
    }

    /**
     * Get outlets by city ID
     * @param cityId - City ID
     * @returns Promise with list of outlets
     */
    static async getOutletsByCity(cityId: number): Promise<ApiResponse<OutletResponse[]>> {
        try {
            const response = await apiFetch(API_ENDPOINTS.OUTLETS_BY_CITY(cityId));
            const data = await response.json();
            return data;
        } catch (error) {
            console.error('Outlets by city API call error:', error);
            throw error;
        }
    }

    /**
     * Get outlets by state ID
     * @param stateId - State ID
     * @returns Promise with list of outlets
     */
    static async getOutletsByState(stateId: number): Promise<ApiResponse<OutletResponse[]>> {
        try {
            const response = await apiFetch(API_ENDPOINTS.OUTLETS_BY_STATE(stateId));
            const data = await response.json();
            return data;
        } catch (error) {
            console.error('Outlets by state API call error:', error);
            throw error;
        }
    }
} 